src/system_identity/iam_client.rs

```diff
@@ -0,0 +1,189 @@
+use base64::{engine::general_purpose, Engine};
+use http::{
+    header::{AUTHORIZATION, CONTENT_TYPE},
+    StatusCode, Uri,
+};
+use serde::{Deserialize, Deserializer};
+use serde_json::json;
+use thiserror::Error;
+
+use crate::{
+    http_client::{HttpClient, HttpClientError},
+    token::AccessToken,
+};
+
+#[derive(Debug, Clone, Error)]
+pub enum IAMClientError {
+    #[error("error creating system identity: `{0}`")]
+    IAMClientError(String),
+}
+
+pub trait IAMClient {
+    fn create_system_identity(
+        &self,
+        token: &AccessToken,
+        pub_key: &[u8],
+    ) -> Result<SystemIdentityCreationResponseData, IAMClientError>;
+}
+
+/// Implementation of the IAMClient trait for a generic HTTP client.
+pub struct HttpIAMClientImpl<'a, C: HttpClient> {
+    http_client: &'a C,
+    name: String,
+    organization_id: String,
+    system_identity_creation_uri: &'a Uri,
+    // other required data? Like access token retriever or auth mechanism? key pair generator?
+}
+
+impl<'a, C: HttpClient> HttpIAMClientImpl<'a, C> {
+    pub fn new(
+        http_client: &'a C,
+        name: String,
+        organization_id: String,
+        system_identity_creation_uri: &'a Uri,
+    ) -> Self {
+        Self {
+            http_client,
+            name,
+            organization_id,
+            system_identity_creation_uri,
+        }
+    }
+
+    fn system_identity_from_token(
+        http_client: &C,
+        name: &str,
+        organization_id: &str,
+        token: &str,    // type
+        pub_key: &[u8], // type and provenance, other modes accepted?
+        system_identity_creation_uri: &Uri,
+    ) -> Result<SystemIdentityCreationResponseData, HttpClientError> {
```

Comment:
We are receiving all fields from `HttpIAMClientImpl`, would it make sense to use a method (by providing `&self`) instead of a static function?

Reply:
Kind of related, `create_system_identity` looks like a wrapper to this function. Can't we just put this code inside `create_system_identity`?

```rust
    fn create_system_identity(
        &self,
        token: &AccessToken,
        pub_key: &[u8],
    ) -> Result<SystemIdentityCreationResponseData, IAMClientError> {
        Self::system_identity_from_token(
            self.http_client,
            self.name.as_str(),
            self.organization_id.as_str(),
            token,
            pub_key,
            self.system_identity_creation_uri,
        )
        .map_err(|e| IAMClientError::IAMClientError(e.to_string()))
    }
```

Reply:
I was thinking in something similar, if we made it a method it would be:

```rust
    fn create_system_identity(
        &self,
        token: &AccessToken,
        pub_key: &[u8],
    ) -> Result<SystemIdentityCreationResponseData, IAMClientError> {
        self.system_identity_from_token(
            token,
            pub_key,
        )
        .map_err(|e| IAMClientError::IAMClientError(e.to_string()))
    }
```

Reply:
Changed in https://github.com/newrelic/newrelic-auth-rs/pull/25/commits/3f58cb9e612c9eeb64dab622dd78586327b78371, but this will change further on next commits.